Building VW statically with Clang does not work

From #1691, @arielf found a regression in Linux build post CMAKE-changes 

> Hi,
> 
> I'd like to reopen this issue. Basically, we have a regression. 
> 
> Since the switch to `cmake`, I have trouble building `vw` statically with clang, on Ubuntu 18.04 by following the (new) instructions.
> 
> Starting with a fresh clone:
> ```
> % git clone git://github.com/VowpalWabbit/vowpal_wabbit.git vwclean
> Cloning into 'vwclean'...
> remote: Enumerating objects: 91, done.
> remote: Counting objects: 100% (91/91), done.
> remote: Compressing objects: 100% (65/65), done.
> remote: Total 57865 (delta 35), reused 51 (delta 25), pack-reused 57774
> Receiving objects: 100% (57865/57865), 120.07 MiB | 10.73 MiB/s, done.
> Resolving deltas: 100% (43117/43117), done.
> ```
> 
> Proceeding with build according to instructions:
> ```
> % cd vwclean
> % mkdir build
> % cd build
> % export CC=clang CXX=clang++
> % cmake ..  -DCMAKE_BUILD_TYPE=Release  -DSTATIC_LINK_VW=On
> -- VowpalWabbit Version: 8.6.1
> -- Number of processors: 8
> -- Boost version: 1.65.1
> -- Found the following Boost libraries:
> --   program_options
> --   system
> --   thread
> --   unit_test_framework
> --   chrono
> --   date_time
> --   atomic
> -- Configuring done
> -- Generating done
> -- Build files have been written to: ~/src/vw/vwclean/build
> 
> % make
> [... long output  trimmed, compilation works, but final linking fails ...]
> 
> [100%] Linking CXX executable vw-unit-test.out
> CMakeFiles/vw-unit-test.out.dir/main.cc.o: In function `main':
> /usr/include/boost/test/unit_test.hpp:63: undefined reference to `boost::unit_test::unit_test_main(bool (*)(), int, char**)'
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> test/unit_test/CMakeFiles/vw-unit-test.out.dir/build.make:230: recipe for target 'test/unit_test/vw-unit-test.out' failed
> make[2]: *** [test/unit_test/vw-unit-test.out] Error 1
> CMakeFiles/Makefile2:821: recipe for target 'test/unit_test/CMakeFiles/vw-unit-test.out.dir/all' failed
> make[1]: *** [test/unit_test/CMakeFiles/vw-unit-test.out.dir/all] Error 2
> Makefile:140: recipe for target 'all' failed
> make: *** [all] Error 2
> ```
> 
> It is possible that `LD=clang++` should be added to environment (?)
> 
> Interestingly, going one level up (not exactly following the instructions), seems to work slightly better:
> ```
> % cd ..
> % make
> mkdir -p build
> cd build; cmake ..
> -- VowpalWabbit Version: 8.6.1
> -- Number of processors: 8
> -- Boost version: 1.65.1
> -- Found the following Boost libraries:
> --   program_options
> --   system
> --   thread
> --   unit_test_framework
> --   chrono
> --   date_time
> --   atomic
> -- Configuring done
> -- Generating done
> -- Build files have been written to: ~/src/vw/vwclean/build
> cd build; make -j8 vw-bin
> make[1]: Entering directory '~/src/vw/vwclean/build'
> make[2]: Entering directory '~/src/vw/vwclean/build'
> make[3]: Entering directory '~/src/vw/vwclean/build'
> make[4]: Entering directory '~/src/vw/vwclean/build'
> make[4]: Leaving directory '~/src/vw/vwclean/build'
> [  4%] Built target allreduce
> make[4]: Entering directory '~/src/vw/vwclean/build'
> make[4]: Leaving directory '~/src/vw/vwclean/build'
> [ 97%] Built target vw
> make[4]: Entering directory '~/src/vw/vwclean/build'
> make[4]: Leaving directory '~/src/vw/vwclean/build'
> [100%] Built target vw-bin
> ...
> ```
> 
> Now we have a `vw` executable, but it crashes:
> ```
> % ./build/vowpalwabbit/vw
> Command terminated by signal 11
> ```
> 
> And `make test` is failing 5 of 5 tests.
> 
> Another related/new issue which didn't exist pre the `cmake` switch, is that `cmake` overwrites the original `Makefile`s so now running `git bisect run` refuses to check-out the bisected commit because that would overwrite modified files.
> 
> To summarize, since the `cmake` switch we have:
> - Possible missing `LD` setting
> - Inability to build statically
> - When linked manually, executable crashes
> - Instructions not working as they used to
> - `git bisect run` now fails to checkout Makefiles
> 
> I'd love to get to the prior state before all the massive changes when I build statically with clang and everything just works.  Can someone help?  Thanks!

This seems like a big enough problem that it should be tracked under its own issue.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -80,10 +80,15 @@ set(Boost_USE_MULTITHREADED ON)
 set(Boost_USE_STATIC_RUNTIME OFF)
 
 set(THREADS_PREFER_PTHREAD_FLAG ON)
+find_package(Threads REQUIRED)
+
+set(LINK_THREADS Threads::Threads)
+if(STATIC_LINK_VW)
+  set(LINK_THREADS -Wl,--whole-archive -lpthread -Wl,--no-whole-archive)
+endif()
 
 # Align and foreach are also required, for some reason they can't be specified as components though.
 find_package(Boost REQUIRED COMPONENTS program_options system thread unit_test_framework)
-find_package(Threads REQUIRED)
 find_package(ZLIB REQUIRED)
 
 add_subdirectory(cluster)

--- a/vowpalwabbit/CMakeLists.txt
+++ b/vowpalwabbit/CMakeLists.txt
@@ -63,7 +63,7 @@ source_group(explore FILES ${explore_all_headers})
 
 add_library(vw ${vw_all_sources} ${vw_all_headers} ${explore_all_headers})
 
-target_link_libraries(vw PUBLIC rapidjson Boost::program_options Threads::Threads ZLIB::ZLIB allreduce ${CMAKE_DL_LIBS})
+target_link_libraries(vw PUBLIC rapidjson Boost::program_options ${LINK_THREADS} ZLIB::ZLIB allreduce ${CMAKE_DL_LIBS})
 
 if(GCOV)
   target_link_libraries(vw PUBLIC gcov)

